### PR TITLE
add-test-coverage-for-security-add-and-security-key

### PR DESCRIPTION
**Pull request by qiaozha**
fix https://github.com/Azure/autorest.typescript/issues/1414

### CODE REVIEW

File: test/rlcIntegration/securityAadRest.spec.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import SecurityAADRestClient from "./generated/securityAADRest/src";
+import { assert } from "chai";
+
+describe("Security AAD Rest Client", () => {
+  it("should be able to instantiate a new client without Typescript errors", () => {
+    const client = SecurityAADRestClient("subscriptionId" as any);
```
Comment (qiaozha) [inline test/rlcIntegration/securityAadRest.spec.ts:6]: I wasn't very sure how to create a fake TokenCredential here.

File: test/rlcIntegration/securityKeyRest.spec.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import SecurityKeyRestClient from "./generated/securityKeyRest/src";
+import { assert } from "chai";
+
+describe("Security AAD Rest Client", () => {
+  it("should be able to instantiate a new client without Typescript errors", () => {
+    const client = SecurityKeyRestClient("subscriptionId" as any);
```
Comment (qiaozha) [inline test/rlcIntegration/securityKeyRest.spec.ts:6]: Same here, I wasn't very sure how to create a fake KeyCredential here.

File: test/rlcIntegration/securityAadRest.spec.ts (hunk 1 above)
Comment (joheredi) [reply test/rlcIntegration/securityAadRest.spec.ts:6]: I think you should be able to pass an object like

{getToken: async () => {assert.equal(validate, something); return { accessToken: "FakeToken"}}}

**Review by joheredi — COMMENTED** [review]
How do we handle anonymous credentials and combination of credentials?

I would think we could do that with overloads?

File: test/rlcIntegration/securityAadRest.spec.ts (hunk 1 above)
Comment (qiaozha) [reply test/rlcIntegration/securityAadRest.spec.ts]: Updated.

File: test/rlcIntegration/securityKeyRest.spec.ts (hunk 2 above)
Comment (qiaozha) [reply test/rlcIntegration/securityKeyRest.spec.ts]: updated

**Comment by qiaozha** [comment]
> How do we handle anonymous credentials and combination of credentials?
> 
> I would think we could do that with overloads?

What exactly is anonymous credentials ? we haven't get the case of combination of credentials swagger in testserver ready yet. but the generated code would look like 
```
export default function createClient(
  credentials: TokenCredential | KeyCredential,
  options: ClientOptions = {}
):
```
in that case, we will write a test to call createClient twice one for TokenCredential and one for KeyCredential.

**Comment by joheredi** [comment]
> > How do we handle anonymous credentials and combination of credentials?
> > I would think we could do that with overloads?
> 
> What exactly is anonymous credentials ? we haven't get the case of combination of credentials swagger in testserver ready yet. but the generated code would look like
> 
> ```
> export default function createClient(
>   credentials: TokenCredential | KeyCredential,
>   options: ClientOptions = {}
> ):
> ```
> 
> in that case, we will write a test to call createClient twice one for TokenCredential and one for KeyCredential.

I think anonymous credentials mean that credentials are optional IIRC the spec allows AnonymousCredential as well. I would imagine for that we'd generate an overload that doesn't take any credentials

**Comment by qiaozha** [comment]
In that case, we only need to set the add-credentials: false, I remember there's already a case for that, let me find one.

**Comment by qiaozha** [comment]
https://github.com/Azure/autorest.typescript/blob/main/test/rlcIntegration/generated/urlRest/src/urlRestClient.ts#L7-L9 here's an example of we set add-credentials: false, and how the client factory function looks like

**Comment by joheredi** [comment]
I'm thinking of the scenario where both TokenCredentials and AnonymousCredentials are defined in the swagger

**Review by joheredi — APPROVED** [review]
If reasonable, please add the Anonymous credential support in this PR. Otherwise, can you file an issue for that and maybe try to prioritize?

**Comment by qiaozha** [comment]
> If reasonable, please add the Anonymous credential support in this PR. Otherwise, can you file an issue for that and maybe try to prioritize?

create an issue here https://github.com/Azure/autorest.typescript/issues/1447 to track the support of anonymous credential

**Pull request merged**